File: Swift/Utils/Utils/DownloadUtils.m
#import "DownloadUtils.h"

@implementation DownloadUtils

-(void) download: (NSString *)url name: (NSString *)name path: (NSString *)path {
    self.fileName = name;
    self.filePath = path;
    self.progress = 0;
    NSURL * downloadUrl = [NSURL URLWithString:url];
    NSURLRequest * req = [NSURLRequest requestWithURL:downloadUrl cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:60];
    NSURLConnection * conn = [NSURLConnection connectionWithRequest:req delegate:self];
    [conn start];
}

-(void) connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    if ([FileUtils fileExists:self.fileName filePath:self.filePath]) {
        [FileUtils appendFile:self.fileName savePath:self.filePath fileData:data];
    } else {
        [FileUtils rewriteFile:self.fileName savePath:self.filePath fileData:data];
    }
    self.progress += data.length;
    if ([self.delegate respondsToSelector:@selector(downloadUtils:progress:)]) {
        [self.delegate downloadUtils:self progress:self.progress];
    }
}

-(void) connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    if ([self.delegate respondsToSelector:@selector(downloadUtils:beginDownload:)]) {
        [self.delegate downloadUtils:self beginDownload:0];
    }
}

-(void) connectionDidFinishLoading:(NSURLConnection *)connection {
    if ([self.delegate respondsToSelector:@selector(downloadUtils:endDownload:filePath:)]) {
        [self.delegate downloadUtils:self endDownload:self.progress filePath:[FileUtils buildPath:self.fileName path:self.filePath]];
    }
}

-(void) connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    if ([self.delegate respondsToSelector:@selector(downloadUtils:error:)]) {
        [self.delegate downloadUtils:self error:error.localizedDescription];
    }
}

-(void) connection:(NSURLConnection *)connection willSendRequestForAuthenticationChallenge:(NSURLAuthenticationChallenge *)challenge {
    NSURLCredential * cred = [NSURLCredential credentialForTrust:challenge.protectionSpace.serverTrust];
    [challenge.sender useCredential:cred forAuthenticationChallenge:challenge];
    [challenge.sender continueWithoutCredentialForAuthenticationChallenge:challenge];
}

@end

File: Swift/Utils/DownloadUtils.swift
import UIKit

@objc
protocol DownloadUtilsDelegate: NSObjectProtocol {
    optional func downloadUtils(downloadUtils: DownloadUtils, beginDownload: Int);
    optional func downloadUtils(downloadUtils: DownloadUtils, progress: Int);
    optional func downloadUtils(downloadUtils: DownloadUtils, endDownload: Int, filePath: String);
    optional func downloadUtils(downloadUtils: DownloadUtils, error: NSString);
}

class DownloadUtils: NSObject, NSURLConnectionDataDelegate, NSURLConnectionDelegate {

    var delegate: DownloadUtilsDelegate?
    var fileName: String?
    var filePath: String?
    var progress: Int?
    
    func download(url: String, name: String, path: String) {
        fileName = name
        filePath = path
        progress = 0
        var downloadUrl = NSURL(string: url)
        var req = NSURLRequest(URL: downloadUrl!, cachePolicy: NSURLRequestCachePolicy.UseProtocolCachePolicy, timeoutInterval: 60)
        var conn = NSURLConnection(request: req, delegate: self)
        conn!.start()
    }
    
    func connection(connection: NSURLConnection, didReceiveData data: NSData) {
        if (FileUtils.fileExists(fileName!, filePath: filePath!)) {
            FileUtils.appendFile(fileName!, savePath: filePath!, fileData: data)
        } else {
            FileUtils.rewriteFile(fileName!, savePath: filePath!, fileData: data)
        }
        progress! += data.length
        self.delegate?.downloadUtils?(self, progress: progress!)
    }
    
    func connection(connection: NSURLConnection, didReceiveResponse response: NSURLResponse) {
        self.delegate?.downloadUtils?(self, beginDownload: 0)
    }
    
    func connectionDidFinishLoading(connection: NSURLConnection) {
        self.delegate?.downloadUtils?(self, endDownload: progress!, filePath: FileUtils.buildPath(fileName!, path: filePath!))
    }
    
    func connection(connection: NSURLConnection, didFailWithError error: NSError) {
        self.delegate?.downloadUtils?(self, error: error.localizedDescription)
    }
    
    func connection(connection: NSURLConnection, canAuthenticateAgainstProtectionSpace protectionSpace: NSURLProtectionSpace) -> Bool {
        return protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust
    }
    
    func connection(connection: NSURLConnection, didReceiveAuthenticationChallenge challenge: NSURLAuthenticationChallenge) {
        if (challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust) {
            challenge.sender.useCredential(NSURLCredential(forTrust: challenge.protectionSpace.serverTrust), forAuthenticationChallenge: challenge)
        }
        challenge.sender.continueWithoutCredentialForAuthenticationChallenge(challenge)
    }
    
}
